chain/beacon/drand/drand.go

```diff
@@ -3,57 +3,200 @@ package drand
 import (
 	"context"
 	"sync"
+	"time"
 
+	"github.com/filecoin-project/lotus/build"
 	"github.com/filecoin-project/lotus/chain/beacon"
 	"github.com/filecoin-project/lotus/chain/types"
 	"github.com/filecoin-project/specs-actors/actors/abi"
+	"golang.org/x/xerrors"
 
+	logging "github.com/ipfs/go-log"
+
+	dbeacon "github.com/drand/drand/beacon"
+	"github.com/drand/drand/core"
+	dkey "github.com/drand/drand/key"
 	dnet "github.com/drand/drand/net"
 	dproto "github.com/drand/drand/protobuf/drand"
 )
 
+var log = logging.Logger("drand")
+
+var drandServers = []string{
+	"drand-test1.nikkolasg.xyz:5001",
+}
+
+var drandPubKey *dkey.DistPublic
+
+func init() {
+	drandPubKey = new(dkey.DistPublic)
+	err := drandPubKey.FromTOML(&dkey.DistPublicTOML{Coefficients: build.DrandCoeffs})
+	if err != nil {
+		panic(err)
+	}
+}
+
+type drandPeer struct {
+	addr string
+	tls  bool
+}
+
+func (dp *drandPeer) Address() string {
+	return dp.addr
+}
+
+func (dp *drandPeer) IsTLS() bool {
+	return dp.tls
+}
+
 type DrandBeacon struct {
 	client dnet.Client
+	peers  []dnet.Peer
+
+	pubkey *dkey.DistPublic
+
+	// seconds
+	interval time.Duration
+
+	drandGenTime uint64
+	filGenTime   uint64
+	filRoundTime uint64
 
 	cacheLk    sync.Mutex
-	localCache map[int64]types.BeaconEntry
+	localCache map[uint64]types.BeaconEntry
 }
 
-func NewDrandBeacon() *DrandBeacon {
-	return &DrandBeacon{
+func NewDrandBeacon(genesisTs, interval uint64) (*DrandBeacon, error) {
+	if genesisTs == 0 {
+		panic("what are you doing this cant be zero")
+	}
+	db := &DrandBeacon{
 		client:     dnet.NewGrpcClient(),
-		localCache: make(map[int64]types.BeaconEntry),
+		localCache: make(map[uint64]types.BeaconEntry),
+	}
+	for _, ds := range drandServers {
+		db.peers = append(db.peers, &drandPeer{addr: ds, tls: true})
```

Comment:
It should be true but just in case, you might want to load that info from the group file no ?